service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/flexibleresource/CloneFlexibleResourceStep.java

```diff
@@ -0,0 +1,129 @@
+  private final ControlledFlexibleResource sourceFlexResource;
+  private final CloningInstructions resolvedCloningInstructions;
+
+  public CloneFlexibleResourceStep(
+      SamService samService,
+      AuthenticatedUserRequest userRequest,
+      ControlledResourceService controlledResourceService,
+      ControlledFlexibleResource sourceFlexResource,
+      CloningInstructions resolvedCloningInstructions) {
+    this.samService = samService;
+    this.userRequest = userRequest;
+    this.controlledResourceService = controlledResourceService;
+    this.sourceFlexResource = sourceFlexResource;
+    this.resolvedCloningInstructions = resolvedCloningInstructions;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    FlightMap inputParameters = flightContext.getInputParameters();
+    FlightMap workingMap = flightContext.getWorkingMap();
+
+    FlightUtils.validateRequiredEntries(
+        inputParameters,
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID,
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID);
+
+    String resourceName =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_NAME,
+            String.class);
+    String description =
+        FlightUtils.getInputParameterOrWorkingValue(
+            flightContext,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_DESCRIPTION,
+            WorkspaceFlightMapKeys.ResourceKeys.PREVIOUS_RESOURCE_DESCRIPTION,
+            String.class);
+
+    UUID destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    UUID destinationResourceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID, UUID.class);
+    UUID destinationFolderId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_FOLDER_ID, UUID.class);
+
+    // New flex resource to be created in flight.
+    ControlledFlexibleResource destinationFlexResource =
+        buildDestinationControlledFlexResource(
+            sourceFlexResource,
+            destinationWorkspaceId,
+            destinationResourceId,
+            destinationFolderId,
+            resourceName,
+            description,
+            samService.getUserEmailFromSamAndRethrowOnInterrupt(userRequest));
+
+    ControlledResourceIamRole iamRole =
+        IamRoleUtils.getIamRoleForAccessScope(sourceFlexResource.getAccessScope());
+
+    ControlledFlexibleResource clonedFlexResource =
+        controlledResourceService
+            .createControlledFlexResourceSyncWithoutCreationParameters(
```

Comment:
this is a very specific method that i don't suspect other call path will be using for creating flex resources. Can you move it here?

Reply:
Yeah, I agree this method is quite specific. However, Dan said it's bad practice to have API objects in the flight.

So, I want to avoid using these here, and handle them in the controller.

Reply:
> However, Dan said it's bad practice to have API objects in the flight.

I want to keep all API objects in the controller. I never want to use them as parameters to a flight; they should never be serialized.

What that means is, we need an internal class for the "creation parameters"; let's call it `FlexResourceCreationParameters`. So you would not pass `body` [here](https://github.com/DataBiosphere/terra-workspace-manager/blob/0121a575a4734c586d845fed394cf7d514488751/service/src/main/java/bio/terra/workspace/app/controller/ControlledFlexibleResourceApiController.java#L85). You would convert the relevant fields of `body` into `FlexResourceCreationParameters` and use that.

I am asking you to do it differently than what we have done so far. I think all of the other controlled resources are passing around the API object as creation parameters. And as you see, one of the issues that creates is it makes implementing clone awkward, because you end up creating the API object.

I _should_ have caught this on the review. Sorry about that.

Reply:
I think you could have a single endpoint in ControlledResourceService that would work for both cases, instead of needing the `createControlledFlexResourceSyncWithoutCreationParameters`